src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -3,39 +3,60 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use anyhow::Result;
+use anyhow::{anyhow, Context, Result};
 
 use crate::{
     hypervisor_persist::HypervisorState, HypervisorConfig, MemoryConfig,
     VcpuThreadIds, VsockDevice, HYPERVISOR_QEMU,
 };
-use kata_types::capabilities::{Capabilities, CapabilityBits};
+use kata_types::{
+    capabilities::{Capabilities, CapabilityBits},
+    config::KATA_PATH,
+};
 use async_trait::async_trait;
+use std::collections::HashMap;
 use persist::sandbox_persist::Persist;
 
 const VSOCK_SCHEME: &str = "vsock";
-const VSOCK_AGENT_CID: u32 = 3;
-const VSOCK_AGENT_PORT: u32 = 1024;
+
 #[derive(Debug)]
 pub struct QemuInner {
+    /// sandbox id
+    id: String,
+
     config: HypervisorConfig,
+    devices: Vec<DeviceType>,
 }
 
 impl QemuInner {
     pub fn new() -> QemuInner {
         QemuInner {
+            id: "".to_string(),
             config: Default::default(),
+            devices: Vec::new(),
         }
     }
 
-    pub(crate) async fn prepare_vm(&mut self, _id: &str, _netns: Option<String>) -> Result<()> {
+    pub(crate) async fn prepare_vm(&mut self, id: &str, _netns: Option<String>) -> Result<()> {
         info!(sl!(), "Preparing QEMU VM");
+
+        self.id = id.to_string();
+
+        self.devices.push(DeviceType::Vsock(
+            VsockDevice::new(self.id.clone())
+                .await
+                .context("qemu: create agent vsock")?,
+        ));
+
         Ok(())
     }
 
     pub(crate) async fn start_vm(&mut self, _timeout: i32) -> Result<()> {
         info!(sl!(), "Starting QEMU VM");
 
+        let vm_path = [KATA_PATH, self.id.as_str()].join("/");
+        std::fs::create_dir_all(vm_path)?;
```

Comment:
Actually yes, you're right, this would better be done in `prepare_vm()`.  Would it be acceptable though do move it in a subseqent PR on the grounds that it does no actual harm and I'd really like get this merged asap to be able to move ahead?

Reply:
I've moved it locally already but I'd prefer to give it some testing (by running with the change for a while).  This will be part of my next PR.